### PR TITLE
[clang][Sema] Make capability narrowing follow normal pointer/int behaviour

### PR DESCRIPTION
**Pull request by jrtc27**
This code dates back to 2017; it's seemingly not needed to get
diagnostics these days, and yields weird, incorrect diagnostics in
real-world cases. If this is still catching something important that we
aren't catching elsewhere then that should be fixed in a more consistent
manner, though I've tried to add tests for important constructs.

Fixes https://github.com/CTSRD-CHERI/llvm-project/issues/483

### CODE REVIEW

File: clang/test/Sema/cheri/init-list-narrowing.c [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+  // hybrid-cxx-error@-2 {{converting capability type 'void * __capability' to non-capability type 'void *' without an explicit cast; if this is intended use __cheri_fromcap}}
+  {(void * __capability)foo},
+  // hybrid-error@-1 {{converting capability type 'void * __capability' to non-capability type 'void *' without an explicit cast; if this is intended use __cheri_fromcap}}
+  {(__intcap)0},
+  // expected-c-warning@-1 {{expression which evaluates to zero treated as a null pointer constant of type 'void *'}}
+  // expected-cxx-error@-2 {{cannot initialize a member subobject of type 'void *' with an rvalue of type '__intcap __attribute__((cheri_no_provenance))'}}
+  {(__intcap)42},
+  // expected-c-warning@-1 {{incompatible integer to pointer conversion initializing 'void *' with an expression of type '__intcap __attribute__((cheri_no_provenance))'}}
+  // expected-cxx-error@-2 {{cannot initialize a member subobject of type 'void *' with an rvalue of type '__intcap __attribute__((cheri_no_provenance))'}}
```
Comment (arichardson) [inline clang/test/Sema/cheri/init-list-narrowing.c:69]: Could you add a few more cases here that should result in a C++ narrowing warning by using an out-of-range constant:
```
extern char foo[];
#define __capability
typedef long __intcap;
constexpr __intcap l = (long)__INT_MAX__ + 1;
constexpr int i = __INT_MIN__;

struct {
  int x;
} ints[] = {
  {(__intcap)(__LONG_MAX__)},
  {(__intcap)&foo[0]},
  {l},
  {(__intcap)i},
};
```
https://godbolt.org/z/7dexja1Gq

**Review by arichardson — COMMENTED** [review]
Been a long time, but I think this was mostly intended to handle `void*__cap -> void*` (since that wasn't handled correctly before. So if we can delete this code now that's great :)

**Comment by arichardson** [comment]
The tests failing in jenkins also all look like they now have better error messages, so this is great.

**Comment by jrtc27** [comment]
I noticed those failing whilst Jenkins was still building, so hit the "pushed before the previous build finished" case that confuses Jenkins and makes it thinks it already built the newer version...

File: clang/test/Sema/cheri/init-list-narrowing.c (hunk 1 above)
Comment (jrtc27) [reply clang/test/Sema/cheri/init-list-narrowing.c:87]: Added a bunch of `(long)__INT_MAX__ + 1` tests; don't particularly want to be exhaustively testing the boundary cases, just that we can get that diagnostic out

**Review by arichardson — APPROVED** [review]
